### DISCUSSION TITLE
Using CN: Common Name and OU: Organization Unit for Certificate Validation

### DISCUSSION THREAD

**Discussion by lxjoyner** [Hakky54/ayza#219] [General]
I have a need to use CN and OU in certificate validation using spring. Currently using spring to connect using SSL but need to add certificate validation to only allow certificates with specific CN and OU. Can I use the one in this comment below and just add the OU. If so what is the pattern syntax for parsing the OU from the certificate.                                                                        

[AdditionalCertificateValidationsAspect.java](https://github.com/Hakky54/java-tutorials/blob/main/spring-cn-validation-with-aop/server-with-cn-validation/src/main/java/nl/altindag/server/aspect/AdditionalCertificateValidationsAspect.java)
[Added CN validator based on AOP](https://github.com/Hakky54/java-tutorials/commit/3bb7bbd249ec6d2375fb1126999a3d010ebf457b)

**Comment by Hakky54** [comment]
Hi @lxjoyner 

You can reuse the code snippet which is present in the [AdditionalCertificateValidationsAspect.java](https://github.com/Hakky54/java-tutorials/blob/main/spring-cn-validation-with-aop/server-with-cn-validation/src/main/java/nl/altindag/server/aspect/AdditionalCertificateValidationsAspect.java)

I would suggest the following refactored snippet:

```java
public class AdditionalCertificateValidationsAspect {

    private static final Pattern COMMON_NAME_PATTERN = Pattern.compile("(?<=CN=)(.*?)(?=,)");
    private static final Pattern ORGANISATION_UNIT_PATTERN = Pattern.compile("(?<=OU=)(.*?)(?=,)");

    // other methods can be found in the original source code

    private Optional<String> getCommonName(String subjectDistinguishedName) {
        return getSubfield(COMMON_NAME_PATTERN, subjectDistinguishedName);
    }

    private Optional<String> getOrganizationUnit(String subjectDistinguishedName) {
        return getSubfield(ORGANISATION_UNIT_PATTERN, subjectDistinguishedName);
    }

    private Optional<String> getSubfield(Pattern pattern, String subjectDistinguishedName) {
        var matcher = pattern.matcher(subjectDistinguishedName);

        if (matcher.find()) {
            return Optional.of(matcher.group());
        } else {
            return Optional.empty();
        }
    }

}
```

**Reply by Hakky54** [reply]
@lxjoyner Did it work for you?
